### PR TITLE
MINIFICPP-2081 - Read windows events from log file

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/windows-event-log/wel/EventPath.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+EventPath::Kind EventPath::kind() const {
+  return kind_;
+}
+
+EVT_QUERY_FLAGS EventPath::getQueryFlags() const {
+  switch (kind_) {
+    case Kind::CHANNEL: return EvtQueryChannelPath;
+    case Kind::FILE: return EvtQueryFilePath;
+    default: gsl_Assert(false);
```
Comment (szaszm) [inline extensions/windows-event-log/wel/EventPath.cpp:56]: Can we omit the default path, or does it result in some kind of error? If we omit it, we should get a warning if we were to ever add a new enum value.

File: extensions/windows-event-log/tests/CWELTestUtils.h [hunk 2]
```diff
@@ -94,3 +94,13 @@ class OutputFormatTestController : public TestController {
   std::string output_format_;
   std::optional<std::string> json_format_;
 };
+
+void generateLogFile(const std::wstring& channel, const std::filesystem::path& path) {
+  HANDLE hEventLog = OpenEventLog(NULL, std::string(channel.begin(), channel.end()).c_str());
+
+  if (!EvtExportLog(NULL, channel.c_str(), L"*", path.wstring().c_str(), EvtExportLogChannelPath)) {
+    throw std::runtime_error("Failed to export logs: " + std::system_category().message(GetLastError()));
```
Comment (szaszm) [inline extensions/windows-event-log/tests/CWELTestUtils.h:102]: We could throw `std::system_error` instead, with the error code inside.
```suggestion
    throw std::system_error{gsl:narrow<int>(GetLastError()), std::system_category(), "Failed to export logs"};
```

File: extensions/windows-event-log/wel/EventPath.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#include "EventPath.h"
+#include <utility>
+#include "utils/StringUtils.h"
+
+namespace org::apache::nifi::minifi::wel {
+
+EventPath::EventPath(std::wstring wstr) : EventPath(std::string(wstr.begin(), wstr.end())) {}
+
+EventPath::EventPath(std::string str) {
+  constexpr std::string_view saved_log_prefix = "SavedLog:";
```
Comment (lordgamez) [inline extensions/windows-event-log/wel/EventPath.cpp:28]: The processor documentation should be updated with this prefix usage

File: extensions/windows-event-log/tests/CWELTestUtils.h [hunk 4]
```diff
@@ -94,3 +94,13 @@ class OutputFormatTestController : public TestController {
   std::string output_format_;
   std::optional<std::string> json_format_;
 };
+
+void generateLogFile(const std::wstring& channel, const std::filesystem::path& path) {
+  HANDLE hEventLog = OpenEventLog(NULL, std::string(channel.begin(), channel.end()).c_str());
```
Comment (lordgamez) [inline extensions/windows-event-log/tests/CWELTestUtils.h:99]: Even though Windows API uses the Hungarian notation for the `HANDLE` object I think we should stick with our style guide in our code for these variables as well, so simply `event_log` would be better.

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp [hunk 5]
```diff
@@ -527,4 +549,24 @@ TEST_CASE("ConsumeWindowsEventLog Simple JSON works with UserData", "[cwel][json
   }
 }
 
+TEST_CASE("ConsumeWindowsEventLog can process events from a log file", "[cwel][logfile]") {
+  std::string event = LogFileTestController{}.run();
+  std::cout << "Event: " << event << std::endl;
```
Comment (lordgamez) [inline extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp:554]: Do we need this `cout` line?

File: extensions/windows-event-log/tests/CWELTestUtils.h (hunk 2 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/tests/CWELTestUtils.h]: done

File: extensions/windows-event-log/wel/EventPath.cpp (hunk 3 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/wel/EventPath.cpp:28]: done

File: extensions/windows-event-log/tests/CWELTestUtils.h (hunk 4 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/tests/CWELTestUtils.h]: renamed

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp [hunk 6]
```diff
@@ -65,6 +65,28 @@ class SimpleFormatTestController : public OutputFormatTestController {
...
+ public:
+  LogFileTestController(): OutputFormatTestController("", "*", "JSON", "Simple") {
+    log_file_ = createTempDirectory() / "cwel-events.evtx";
+    channel_ = "SavedLog:" + log_file_.string();
+  }
+
+ protected:
+  void dispatchBookmarkEvent() override {
+    reportEvent(APPLICATION_CHANNEL, "Event zero from file: this is in the past");
+    generateLogFile(std::wstring{APPLICATION_CHANNEL.begin(), APPLICATION_CHANNEL.end()}, log_file_);
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp:78]: `APPLICATION_CHANNEL` is `std::wstring` in `BookmarkTests.cpp` and `std::string` in this file, there are also two different `reportEvent` implementations, this should be revisited to unify these

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp (hunk 5 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp]: removed

File: extensions/windows-event-log/tests/CWELTestUtils.h [hunk 7]
```diff
@@ -94,3 +94,13 @@ class OutputFormatTestController : public TestController {
   std::string output_format_;
   std::optional<std::string> json_format_;
 };
+
+void generateLogFile(const std::wstring& channel, const std::filesystem::path& path) {
+  HANDLE event_log = OpenEventLog(NULL, std::string(channel.begin(), channel.end()).c_str());
```
Comment (szaszm) [inline extensions/windows-event-log/tests/CWELTestUtils.h:99]: We should use RAII to close the handle. (`unique_ptr` with deleter, or `gsl::finally`)
Now it's leaked when we throw.

File: extensions/windows-event-log/wel/EventPath.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    FILE
+  };
+
+  EventPath() = default;
+  explicit EventPath(std::wstring wstr);
+  explicit EventPath(std::string str);
+  const std::wstring& wstr() const;
+  const std::string& str() const;
+  EVT_QUERY_FLAGS getQueryFlags() const;
+  Kind kind() const;
```
Comment (szaszm) [inline extensions/windows-event-log/wel/EventPath.h:40]: Consider using `noexcept` for these. We should use it more often in cases where it's safe. Same with `constexpr`: I think all member functions could be `constexpr`, we just can't have a heap allocated object cross from compile-time to runtime.

File: extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/windows-event-log/tests/ConsumeWindowsEventLogTests.cpp:78]: Oh I see now, thanks for clarifying! You are right, it should be unified.

File: extensions/windows-event-log/tests/CWELTestUtils.h (hunk 7 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/tests/CWELTestUtils.h:100]: done

File: extensions/windows-event-log/wel/EventPath.h (hunk 8 above)
Comment (adamdebreceni) [reply extensions/windows-event-log/wel/EventPath.h]: done

**Review by lordgamez — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]

**Pull request closed without merging**